src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

```diff
@@ -0,0 +1,214 @@
+
+        private PollingCounter? _lookupsRequestedCounter;
+        private EventCounter? _lookupsDuration;
+
+        private long _lookupsRequested;
+
+        protected override void OnEventCommand(EventCommandEventArgs command)
+        {
+            if (command.Command == EventCommand.Enable)
+            {
+                // The cumulative number of name resolution requests started since the process started.
+                _lookupsRequestedCounter ??= new PollingCounter("dns-lookups-requested", this, () => Interlocked.Read(ref _lookupsRequested))
+                {
+                    DisplayName = "DNS Lookups Requested"
+                };
+
+                _lookupsDuration ??= new EventCounter("dns-lookups-duration", this)
+                {
+                    DisplayName = "Average DNS Lookup Duration",
+                    DisplayUnits = "ms"
+                };
+            }
+        }
+
+        private const int MaxIPFormattedLength = 128;
+
+        [Event(ResolutionStartEventId, Level = EventLevel.Informational)]
+        public ValueStopwatch ResolutionStart(string hostNameOrAddress)
+        {
+            Debug.Assert(hostNameOrAddress != null);
+
+            if (IsEnabled())
+            {
+                Interlocked.Increment(ref _lookupsRequested);
+                WriteEvent(ResolutionStartEventId, hostNameOrAddress);
+                return ValueStopwatch.StartNew();
+            }
+
+            return default;
+        }
+
+        [NonEvent]
+        public ValueStopwatch ResolutionStart(IPAddress address)
+        {
+            Debug.Assert(address != null);
+
+            if (IsEnabled())
+            {
+                Interlocked.Increment(ref _lookupsRequested);
+                WriteEvent(ResolutionStartEventId, FormatIPAddressNullTerminated(address, stackalloc char[MaxIPFormattedLength]));
+                return ValueStopwatch.StartNew();
+            }
+
+            return default;
+        }
+
+        [NonEvent]
+        public void AfterResolution(string hostNameOrAddress, ValueStopwatch stopwatch, bool successful)
+        {
+            if (stopwatch.IsActive)
+            {
+                double duration = stopwatch.GetElapsedTime().TotalMilliseconds;
+
+                _lookupsDuration!.WriteMetric(duration);
+
+                if (successful)
+                {
+                    ResolutionSuccess(hostNameOrAddress, duration);
+                }
+                else
+                {
+                    ResolutionFailure(hostNameOrAddress, duration);
+                }
+            }
+        }
+
+        [NonEvent]
+        public void AfterResolution(IPAddress address, ValueStopwatch stopwatch, bool successful)
+        {
+            if (stopwatch.IsActive)
+            {
+                double duration = stopwatch.GetElapsedTime().TotalMilliseconds;
+
+                _lookupsDuration!.WriteMetric(duration);
+
+                WriteEvent(
```

Comment:
Same thing here.  You should have this call the `[Event]` methods directly.  The other benefit of doing this is that in the future, when someone adds a new field to the event, you won't have a mismatch here.

Reply:
Based on the conversation above https://github.com/dotnet/runtime/pull/38409#discussion_r447069505, I am marking this as resolved as we can not call into `[Event]` methods directly without the overhead of an extra string allocation.